feat(backend): added timing task check to backend

zhouzilong2020 opened the pull request.

zhouzilong2020 commented:
PTAL

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,74 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+				return nil, err
+			}
+		} else {
+			switch isPassedTimingTaskCheck {
```
true / false doesn't need to use switch

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,74 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
 func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *api.Task, creatorID int, skipIfAlreadyTerminated bool) (*api.Task, error) {
+	// All task should pass timing task check.
+	// Since Time is an auto-increment value, a task would generate a timing task check every time before executing.
+	// However, this is an annoying behavior, and we adopt the following logic to ameliorate this problem.
```
should be explicit about the exact logic instead of asking reader to understand the code. And also for this particular case, it's actually pretty tricky. The way the new logic is written is we would change the task run status, which is actually pretty surprising, given that method only expects to schedule a check (see the method comment)

also ameliorate is an uncommon word, we should choose simple word, e.g. ease

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,74 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+	// However, this is an annoying behavior, and we adopt the following logic to ameliorate this problem.
+	{
+		// fetching history task check
+		var err error
```
We can have a short circuit logic here to skip the check if task.NotBeforeTs is 0, this is more consistent with we set its default value to 0.

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,74 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+				break
+			}
+		}
+		isPassedTimingTaskCheck, err := s.server.passCheck(ctx, s.server, task, api.TaskCheckGeneralEarliestAllowedTime)
```
should this put inside the else block below instead of putting it before if timingTaskCheck == nil block?

tianzhou reviewed (changes requested):
Needs more work

zhouzilong2020 commented:
added some annotation and refactored the code for readability
PTAL @tianzhou

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,69 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+		// All task should pass timing task check.
+		// Since Time is an auto-increment value, a task would generate a timing task check every time before executing.
+		// However, this is an annoying behavior, and we adopt the following logic to ease this problem:
+		// 		1. if no timing check has been schedule yet, schedule one
```
has been scheduled yet

zhouzilong2020 replied:
done

tianzhou commented on server/task_scheduler.go:
```diff
@@ -233,17 +233,28 @@ func (s *TaskScheduler) Register(taskType string, executor TaskExecutor) {
 // ScheduleIfNeeded schedules the task if its required check does not contain error in the latest run
 func (s *TaskScheduler) ScheduleIfNeeded(ctx context.Context, task *api.Task) (*api.Task, error) {
-	// For now, only schema update task has required task check
+	// every task should pass timing task check
```
IIUC,  we would never pass the task check if task.NotBeforeTs == 0. In this case, we won't run the check, and since we won't run the check, passCheck always returns false.

zhouzilong2020 replied:
My fault, forget to add circuit logic to the scheduler

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,69 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+			return nil, err
+		}
+		if isTimingTaskCheckPassed {
+			// Since we allowed user to modify 'notBeforeTs' after creating a task, it is possible previous passed timing check would fail this time.
```
The comment could be more explicit.

Since we allow user to modify 'notBeforeTs' after creating a task, and if the user did change the time to a later time, then we need to schedule another check to block the run.

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -179,6 +179,69 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+		if err != nil {
+			return nil, err
+		}
+		if isTimingTaskCheckPassed {
```
Looks like the block can be simplified.

There are 2 variations.

1. Whether we need to run the check

2. If we need to run the check, whether we need to pass SkipIfAlreadyTerminated

Would be more readable to structrure and comment code in such way.

tianzhou reviewed (changes requested):
Needs more work

zhouzilong2020 commented:
PTAL, added a new payload for timing check.

tianzhou commented on server/task_check_executor_timing.go:
```diff
@@ -0,0 +1,55 @@
+			{
+				Status:  api.TaskCheckStatusError,
+				Code:    common.TaskTimingNotAllowed,
+				Title:   "TimingNotAllowed",
```
Not ready to run

tianzhou commented on server/task_check_executor_timing.go:
```diff
@@ -0,0 +1,55 @@
+				Status:  api.TaskCheckStatusError,
+				Code:    common.TaskTimingNotAllowed,
+				Title:   "TimingNotAllowed",
+				Content: fmt.Sprintf("Did not pass the earliest execution timing: %s", time.Unix(task.NotBeforeTs, 0).Format(dataFormat)),
```
Need to wait until the configured earliest running time: %s

This message is passed to UI for now, so let's make it more readable

tianzhou commented on server/task_check_executor_timing.go:
```diff
@@ -0,0 +1,55 @@
+				Status:  api.TaskCheckStatusSuccess,
+				Code:    common.Ok,
+				Title:   "OK",
+				Content: fmt.Sprintf("Passed the earliest execution timing: %s", time.Unix(task.NotBeforeTs, 0).Format(dataFormat)),
```
Passed the configured earliest running time: %s

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,60 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
 	s.executors[taskType] = executor
 }
 
+func (s *TaskCheckScheduler) shouldScheduleTimingTaskCheck(ctx context.Context, task *api.Task) bool {
```
Should add a comment to explain when we would return true

tianzhou replied:
Also we should return the err instead of failing silently

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,60 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+		return false
+	}
+
+	statusList := []api.TaskCheckRunStatus{api.TaskCheckRunDone, api.TaskCheckRunFailed}
```
what if there is already a running timing check?

tianzhou commented on server/task_check_executor_timing.go:
```diff
@@ -0,0 +1,55 @@
+				Content: fmt.Sprintf("Did not pass the earliest execution timing: %s", time.Unix(task.NotBeforeTs, 0).Format(dataFormat)),
+			},
+		}, nil
+	} else {
```
nit: we do not need else here

tianzhou reviewed (changes requested):
Needs more work

zhouzilong2020 commented:
PTAL

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,77 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
 }
 
+// shouldScheduleTimingTaskCheck will return whether should we schedule a timing task check for current task
+// would return true when:
```
Need to be explicit about if we need to meet any of the following conditions or all of the conditions.

The concurrent structure is missing because IIUC, the logic is 1 && ( 2 || 3)

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,77 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+			if err != nil {
+				return nil, err
+			}
+			return task, nil
```
should not return here otherwise we'd never schedule TaskDatabaseSchemaUpdate

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,77 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+				Payload:                 string(taskCheckPayload),
+				TaskID:                  task.ID,
+				Type:                    api.TaskCheckGeneralEarliestAllowedTime,
+				SkipIfAlreadyTerminated: false,
```
I think we should still somehow respect skipIfAlreadyTerminated, we may move the logic taskCheckPayload.NotBeforeTs != task.NotBeforeT here to determine

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,77 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+		}
+
+		if flag {
+			taskCheckPayload, err := json.Marshal(task.NotBeforeTs)
```
I am not sure this marshal works. We normally marshal a struct instead of marshal a field

tianzhou reviewed (changes requested):
Needs more work

zhouzilong2020 commented:
PTAL.
BTW, I have tested this code locally, and it behaves as expected.

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,90 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+// the logic goes like the following:
+// 1. user specified the notBeforeTs field, hence a timing gate is added
+// 2. there is no running timing task check
+// 3(a). the notBeforeTs has altered since last check (either creation or patch)
```
the notBeforeTs state has been altered since last check (either via new creation or patch).

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,90 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+// 1. user specified the notBeforeTs field, hence a timing gate is added
+// 2. there is no running timing task check
+// 3(a). the notBeforeTs has altered since last check (either creation or patch)
+// 3(b). the notBeforeTs has NOT altered since last check && last check had failed && notBeforeTs has passed now
```
the noteBeforeTs has NOT been altered since last check, however the last check had failed && notBeforeTs has passed now.

zhouzilong2020 commented:
PTAL,
1. Fixed the circuit logic and added a force logic in `shouldScheduleTimingTaskCheck`
2. timing task check will not block scheduling other task checks

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,90 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
 
+// shouldScheduleTimingTaskCheck will return whether should we schedule a timing task check for current task
+// the logic goes like the following:
+// 1. user specified the notBeforeTs field, hence a timing gate is added
```
should update the comment as well.

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -177,8 +177,90 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
+	// the following block is for timing task check
+	{
+		// we only set skipIfAlreadyTerminated to false when user explicitly want to reschedule a taskCheck
+		flag, err := s.shouldScheduleTimingTaskCheck(ctx, task, !skipIfAlreadyTerminated)
```
nit: to make it more readable, usually one would write

forceSchedule := skipIfAlreadyTerminated
flag, err := s.shouldScheduleTimingTaskCheck(ctx, task, forceSchedule)

or 

flag, err := s.shouldScheduleTimingTaskCheck(ctx, task, !skipIfAlreadyTerminated /* forceSchedule */)

tianzhou commented on api/task_check_run.go:
```diff
@@ -77,8 +77,15 @@ const (
 	// TaskCheckInstanceMigrationSchema is the task check type for migrating schemas.
 	TaskCheckInstanceMigrationSchema TaskCheckType = "bb.task-check.instance.migration-schema"
+	// TaskCheckGeneralEarliestAllowedTime is the task check type for earliest allowed time.
+	TaskCheckGeneralEarliestAllowedTime TaskCheckType = "bb.task-check.general.earliest-allowed-time"
```
Per offline discussion, let's update notBeforeTs to earliestAllowedTs.  Negative term is not that readable.

tianzhou reviewed (dismissed):
Overall logic LGTM. Some style comment

zhouzilong2020 commented:
1. correct the function's annotation and changed the position of the `finding issue logic` @dc22413
2. changed the notaion of `NotBeforeTs`, `notBeforeTs` and `not_before_ts` to `EarliestAllowedTs`, `earliestAllowedTs` and `earliest_allowed_ts`  respectively @35f8b91
PTAL

tianzhou commented on store/migration/10001__init_schema.sql:
```diff
@@ -842,14 +842,14 @@ CREATE TABLE task (
 CREATE INDEX idx_task_status ON task(`status`);
 
-CREATE INDEX idx_task_not_before_ts ON task(not_before_ts);
+CREATE INDEX idx_task_not_before_ts ON task(earliest_allowed_ts);
```
still need to change index name

zhouzilong2020 replied:
Did not notice the index name, my bad...
😭

tianzhou reviewed (changes requested):
One leftover

tianzhou reviewed (approved):
LGTM
